fix: Missing bucket name check

AndyOverLord opened the pull request:
<!--

Some key notes before you open a PR:

 1. Select which branch should this PR be merged in?
 2. PR name follows [convention](http://karma-runner.github.io/4.0/dev/git-commit-msg.html)
 3. All tests pass locally, UI and Unit tests
 4. All business logic and validations must be on the server-side
 5. Update necessary Documentation
 6. Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes


Also, if you're new here

- Documentation Guidelines => https://github.com/frappe/erpnext/wiki/Updating-Documentation

- Contribution Guide => https://github.com/frappe/frappe/blob/develop/.github/CONTRIBUTING.md

- Pull Request Checklist => https://github.com/frappe/erpnext/wiki/Pull-Request-Checklist

-->

> Please provide enough information so that others can review your pull request:

Problem: If the user changed the backup limit, then press save. The validate function will give an error message that the bucket name already exists. It would be inconvenient for the user to use a different bucket name to save any changes.

<!-- You can skip this if you're fixing a typo or updating existing documentation -->

> Explain the **details** for making this change. What existing problem does the pull request solve?

So I implemented a flag ''bucket_name_exist'' to indicate if the bucket name exists, if not, then go to the flow of trying to create a bucket.
<!-- Example: When "Adding a function to do X", explain why it is necessary to have a way to do X. -->


<!-- Add images/recordings to better visualize the change: expected/current behviour -->

AndyOverLord committed:
fix: missing bucket name check when user make changes on s3 backup setting

Problem: If the user changed the backup limit, then press save. The validate function will give an error message that the bucket name already exists. It would be inconvenient for the user to use a different bucket name to save any changes.

So I implemented a flag ''bucket_name_exist'' to indicate if the bucket name exists, if not, then go to the flow of trying to create a bucket.

AndyOverLord committed:
fix: extra line removed

AndyOverLord commented:
Hi @Mangesh-Khairnar,

I saw you have been assigned to this PR review, just want to make sure if you need further information about this. Please let me know, thanks.

AOL

Thunderbottom commented:
Hey, "Bucket name already exists" could also mean that the bucket ID has been taken up by someone, not necessary that the user owns the bucket. So I think even if this solution is something that works, it won't work 100% of the times.

Mangesh-Khairnar commented:
@AndyOverLord use [head_bucket](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.head_bucket), this makes sure that the bucket exists and also, checks if user has appropriate access.

AndyOverLord committed:
fix: Use head_bucket Boto3 API 

1. Head_bucket will return 200 Ok if the bucket exists and you have permission to access it.
2. Error code 403 Forbidden, Error code 404 Not Found.
3. Use bucket_name_exist to check if need to create bucket.

Thanks @Mangesh-Khairnar suggestions.
Reference: 
1. https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.head_bucket
2. https://boto3.amazonaws.com/v1/documentation/api/latest/guide/migrations3.html#accessing-a-bucket

Further improvements:
1. Use 'GET' Requestion to check if the Access Key ID and Secret Access Key is valid (Because AWS tier has 20,000 GET requests while only 2,000 LIST requests)
2. Edge case on Error prompts when head_bucket returns an error code other than '403' or '404'.

AndyOverLord committed:
fix: add an extra line to align with original code style

AndyOverLord commented:
@Mangesh-Khairnar, thank you for your suggestions. Please see my commit message for your reference.

Thunderbottom commented on frappe/integrations/doctype/s3_backup_settings/s3_backup_settings.py:
```diff
@@ -34,11 +34,27 @@ def validate(self):
+				frappe.throw(_(f"403 Forbidden. Do not have permission to access this bucket: {bucket_lower}"))
+			elif error_code == '404':
+				# Bucket not found, set bucket_name_exist flag to False to try create bucket
+				bucket_name_exist = False
```
instead of having a flag for checking if bucket name exists, we can actually just create the bucket here and save some lines of code.
```suggestion
				conn.create_bucket(Bucket=bucket_lower, CreateBucketConfiguration={
					'LocationConstraint': self.region})
```

there's also no need to throw:

```python
except ClientError:
	frappe.throw(_("Unable to create bucket: {0}. Change it to a more unique name.").format(bucket_lower))
```

since `head_bucket()` should already tell you if you have access to the bucket if it exists. so the exception is pretty redundant now.

AndyOverLord committed:
Update frappe/integrations/doctype/s3_backup_settings/s3_backup_settings.py

Co-Authored-By: Chinmay Pai <chinmaydpai@gmail.com>

AndyOverLord committed:
fix: remove the flag and redundant exception

When we got a 404 error, we can just create the bucket as 1. The bucket name does not exist in the current bucket. 2. we have permission to access the bucket.

AndyOverLord commented:
Hi @Thunderbottom,
Thank you very much for the code improvements. I have made the changes according to your suggestion.

hrwX commented on frappe/integrations/doctype/s3_backup_settings/s3_backup_settings.py:
```diff
@@ -35,11 +35,20 @@ def validate(self):
+			# If a client error is thrown, then check that it was a 403 error or 404 error.
+			error_code = e.response['Error']['Code']
+			if error_code == '403':
+				frappe.throw(_(f"403 Forbidden. Do not have permission to access this bucket: {bucket_lower}"))
```
```suggestion
				frappe.throw(_("403 Forbidden. Do not have permission to access {0} bucket.").format(bucket_lower))
```

AndyOverLord replied:
Hi hrwX,

Could you please let me know the reason why we use `str.format()` instead of `f"{}"`.
To be backwards compatible with earlier Python versions?

Thank you.

Thunderbottom replied:
Andy, it isn't really necessary that frappe will always be run with Python3. And since Literal String Interpolation wasn't really introduced up until [PEP-0498](https://www.python.org/dev/peps/pep-0498), which got put in Python3.6, this code won't work on Python versions lower than that.

AndyOverLord replied:
Got it, thanks for the explanation.

Thunderbottom commented on frappe/integrations/doctype/s3_backup_settings/s3_backup_settings.py:
```diff
@@ -35,11 +35,20 @@ def validate(self):
-
+			response = conn.head_bucket(Bucket=bucket_lower)
+			# The operation returns a 200 OK if the bucket exists and you have permission to access it.
+			bucket_name_exist = True
```
This flag isn't needed anymore
```suggestion
```

AndyOverLord committed:
Update frappe/integrations/doctype/s3_backup_settings/s3_backup_settings.py

Co-Authored-By: Himanshu <himanshuwarekar@yahoo.com>

AndyOverLord committed:
Update frappe/integrations/doctype/s3_backup_settings/s3_backup_settings.py

Co-Authored-By: Chinmay Pai <chinmaydpai@gmail.com>

AndyOverLord committed:
fix: missing handling the error code '400' - 'Bad Request'

AndyOverLord commented:
Hi @Thunderbottom,

Thanks for the suggestion. When I did testing, I found that when enter non-exist bucket name, AWS return an Error code '400' - bad request instead of '404' Not Found.

Hence I will split all the error messages to 2 parts:
1. if it is '403', it means don't have access to the bucket.
2. if it is other '4xx', simply create a new bucket, as we definitely have access the S3 bucket as tested by `conn.list_buckets()`

AndyOverLord committed:
fix: applied with DeepSource analysis

1. Lines too long
2. Doc Lines too long
3. Expected 2 blank lines between class and method
4. Unused variable removed

AndyOverLord commented:
Hi, 

Any updates about this one?

Thunderbottom commented:
Andy, seems good to me. I'll test it once and merge by tomorrow.

barredterra commented:
@Mergifyio backport version-12-hotfix
